README.md
[![Actions](https://github.com/fluentc/i18n-fluentc/workflows/node/badge.svg)](https://github.com/fluentc/i18n-fluentc/actions?query=workflow%3Anode)
[![Travis](https://img.shields.io/travis/fluentc/i18n-fluentc/master.svg?style=flat-square)](https://travis-ci.org/fluentc/i18n-fluentc)
[![npm version](https://img.shields.io/npm/v/i18n-fluentc.svg?style=flat-square)](https://www.npmjs.com/package/i18n-fluentc)

This is an [i18next backend plugin](https://www.i18next.com/principles/plugins) to be used for [fluentc](https://fluentc.io) service.

If you're not familiar with i18next and how i18next backend plugins works, please first have a look at the [i18next documentation](https://www.i18next.com/how-to/add-or-load-translations#load-using-a-backend-plugin).

<div align="center">
  <a href="https://dashboard.fluentc.io/" target="_blank">
    <img src="https://i.postimg.cc/7PnyZd13/fluentc-logo.png" height="60px">
  </a>
  <img src="https://i.postimg.cc/Cdy09BWb/plus.png" height="60px">
  <a href="https://www.i18next.com/" target="_blank">
    <img src="https://i.postimg.cc/fTygQr8T/i18next-logo.png" height="60px">
  </a>
</div>

<p style="margin-top: 50px;"> </p>

# Obtaining the environmentId from the [dashboard](https://dashboard.fluentc.io)
To use this fluentc-i18next-backend, it's necessary to retrieve the environmentId associated with your project.

Follow these steps:

- Log in to your account on the [fluentc dashboard](https://dashboard.fluentc.io).
- Navigate to the Applications page.
- Select the environment that you wish to use with this module.
- Locate the environmentId displayed within the environment's details section.
- Click on the copy icon to copy the environmentId to your clipboard.

Once you have obtained the environmentId, you can use it with fluentc-i18next-backend to interact with your project.


# Troubleshooting

Make sure you set the `debug` option of i18next to `true`. This will maybe log more information in the developer console.


**Loading translations not working**

Make sure you're using the correct environmentId.


```javascript
import i18next from "i18next";
import Fluentc from "i18n-fluentc";

i18next.use(Fluentc).init({
  backend: {
    environmentId: "[ENVIRONMENTID]",
    referenceLng: "en"
  }
});
```


# Getting started

Source can be loaded via [npm](https://www.npmjs.com/package/i18n-fluentc), `yarn`, `bower` or [downloaded](https://cdn.rawgit.com/fluentc/i18n-fluentc/master/i18nFluentc.min.js) from this repo.

```bash
# npm package
$ npm install i18n-fluentc

# yarn
$ yarn add i18n-fluentc

# bower
$ bower install i18n-fluentc
```

Wiring up:

```js
import i18next from 'i18next';
import Fluentc from 'i18n-fluentc';
// or
const i18next = require('i18next');
const Fluentc = require('i18n-fluentc');

i18next.use(Fluentc).init(i18nextOptions);
```

for Deno:

```js
import i18next from 'https://deno.land/x/i18next/index.js'
import Backend from 'https://deno.land/x/i18n_fluentc/index.js'

i18next.use(Backend).init(i18nextOptions);
```

- As with all modules you can either pass the constructor function (class) to the i18next.use or a concrete instance.
- If you don't use a module loader it will be added to `window.i18nFluentc`

## Backend Options

**IMPORTANT** make sure you do not add your apiKey in the production build to avoid misuse by strangers

```js
{
  // the id of your Fluentc environment
  environmentId: "[ENVIRONMENTID]",

  // the reference language of your project
  referenceLng: '[LNG]',

  // can be used to reload resources in a specific interval (useful in server environments)
  reloadInterval: typeof window !== 'undefined' ? false : 60 * 60 * 1000,
}
```

To load translations only `environmentId` needs to be filled.

Options can be passed in:

**preferred** - by setting options.backend in i18next.init:

```js
import i18next from "i18next";
import Fluentc from "i18n-fluentc";

i18next.use(Fluentc).init({
  backend: options
});
```

on construction:

```js
import Fluentc from "i18n-fluentc";
const fluentc = new Fluentc(options);
```

via calling init:

```js
import Fluentc from "i18n-fluentc";
const fluentc = new Fluentc();
fluentc.init(options);
```

## Additional API endpoints

### backend.getLanguages

Will return a list of all languages in your environment. Languages are available in English and in native language.

```js
import Fluentc from "i18n-fluentc";
const fluentc = new Fluentc(options);

fluentc.getLanguages((err, data) => {
  /*
  data is:

  [
    {
      "label": "Afrikaans",
      "code": "af",
      "localLabel": "Afrikaans"
    },
    {
      "label": "Spanish",
      "code": "es",
      "localLabel": "Español"
    }
    {
      "label": "Japanese",
      "code": "ja",
      "localLabel": "日本語"
    }
  ]
  */
});

// or
const data = await fluentc.getLanguages();

// or
i18next.services.backendConnector.backend.getLanguages(callback);

// or
const data = await i18next.services.backendConnector.backend.getLanguages();
```


## SPECIAL - let the backend determine some options to improve loading

You can load some information from the backend to eg. set supportedLngs for i18next just supporting languages you got in your fluentc environment.

You will get i18next options for (same as above backend.getOptions):

- fallbackLng
- supportedLngs
- load

```js
import i18next from "i18next";
import Fluentc from "i18n-fluentc";

const fluentc = new Fluentc(
  {
    environmentId: "[environmentId]",
    apiKey: "[APIKEY]",
    // referenceLng -> not needed as will be loaded from API
  },
  (err, opts, lngs) => {
    i18next.use(fluentc).init({ ...opts, ...yourOptions }); // yourOptions should not include backendOptions!
  }
);
```

### Special usage with react-i18next without using Suspense

Use `setI18n` to pass in the i18next instance before initializing:

```js
import i18n from "i18next";
import { initReactI18next, setI18n } from "react-i18next";
import FluentcBackend from "i18n-fluentc";

const backendOptions = {
  environmentId: "1d0aa5aa-4660-4154-b6d9-907dbef10bb3"
};

const yourOptions = {
  debug: true,
  interpolation: {
    escapeValue: false
  },
  react: {
    useSuspense: false
  }
};

// this is only used if not using suspense
i18n.options.react = yourOptions.react;
setI18n(i18n);

const backend = new FluentcBackend(backendOptions, (err, opts) => {
  if (err) return console.error(err);
  i18n
    .use(backend)
    // .use(initReactI18next) // keep this if using suspense
    // yourOptions should not include backendOptions!
    .init({ ...opts, ...yourOptions }, (err, t) => {
      if (err) return console.error(err);
    });
});

export default i18n;
```

## TypeScript

To properly type the backend options, you can import the `FluentcBackendOptions` interface and use it as a generic type parameter to the i18next's `init` method, e.g.:

```ts
import i18n from 'i18next'
import FluentcBackend, { FluentcBackendOptions } from 'i18n-fluentc'

i18n
  .use(FluentcBackend)
  .init<FluentcBackendOptions>({
    backend: {
      // fluentc backend options
    },

    // other i18next options
  })
```

<p align="center">
  <a href="https://dashboard.fluentc.io/" target="_blank">
    <img src="https://i.postimg.cc/9MTyH8zy/logo.png" width="240px">
  </a>
</p>

CHANGELOG.md
### 2.1.1

- regression fix for saveMissing [signature](https://github.com/i18next/i18next/issues/1890)

### 2.1.0

- typescript: export the backend options type [105](https://github.com/fluentc/i18n-fluentc/pull/105)

### 2.0.2

- typescript: static type prop

### 2.0.1

- fix if url starts with file:// [100](https://github.com/fluentc/i18n-fluentc/issues/100)

### 2.0.0

- typescript: update for major i18next version



lib/index.js
import { defaults, defer, isMissingOption } from './utils.js'
import { getContent, getAvailableLanguages } from './query.js'
import request from './request.js'

const getDefaults = () => {
  return {
    apiPath: 'https://api.fluentc.io/graphql',
    apiKey: 'da2-wtkl5bpofjbu5ex5iugu4o2mbm',
    referenceLng: 'en',
    reloadInterval: typeof window !== 'undefined' ? false : 60 * 60 * 1000,
    crossDomain: false, // used for XmlHttpRequest
    allowedAddOrUpdateHosts: ['localhost']
  }
}

class Backend {
  constructor (services, options = {}, allOptions = {}, callback) {
    this.services = services
    this.options = options
    this.allOptions = allOptions
    this.type = 'backend'
    if (services && services.environmentId) {
      this.init(null, services, allOptions, options)
    } else {
      this.init(services, options, allOptions, callback)
    }
  }

  init (services, options = {}, allOptions = {}, callback) {
    if (!options.referenceLng && allOptions.fallbackLng && Array.isArray(allOptions.fallbackLng) && allOptions.fallbackLng[0] !== 'dev') {
      options.referenceLng = allOptions.fallbackLng[0]
    }
    this.services = services
    const defOpt = getDefaults()
    const passedOpt = defaults(options, this.options || {})
    if (passedOpt.reloadInterval && passedOpt.reloadInterval < (5 * 60 * 1000)) {
      console.warn('Your configured reloadInterval option is to low.')
      passedOpt.reloadInterval = defOpt.reloadInterval
    }
    this.options = defaults(options, this.options || {}, defOpt)
    this.allOptions = allOptions
    this.somethingLoaded = false
    this.isProjectNotExisting = false

    const hostname = typeof window !== 'undefined' && window.location && window.location.hostname
    if (hostname) {
      this.isAddOrUpdateAllowed = typeof this.options.allowedAddOrUpdateHosts === 'function' ? this.options.allowedAddOrUpdateHosts(hostname) : this.options.allowedAddOrUpdateHosts.indexOf(hostname) > -1

      if (services && services.logger && (allOptions.saveMissing || allOptions.updateMissing)) {
        if (!this.isAddOrUpdateAllowed) {
          services.logger.warn(
            typeof this.options.allowedAddOrUpdateHosts === 'function'
              ? `i18n-fluentc: will not save or update missings because allowedAddOrUpdateHosts returned false for the host "${hostname}".`
              : `i18n-fluentc: will not save or update missings because the host "${hostname}" was not in the list of allowedAddOrUpdateHosts: ${this.options.allowedAddOrUpdateHosts.join(
                ', '
              )} (matches need to be exact).`
          )
        } else if (hostname !== 'localhost') {
          services.logger.warn(`i18n-fluentc: you are using the save or update missings feature from this host "${hostname}".\nMake sure you will not use it in production!`)
        }
      }
    } else {
      this.isAddOrUpdateAllowed = true
    }

    // if (typeof callback === 'function') {
    //   this.getOptions((err, opts, languages) => {
    //     if (err) return callback(err)
    //     this.options.referenceLng = options.referenceLng || opts.referenceLng || this.options.referenceLng
    //     callback(null, opts, languages)
    //   })
    // }

    if (this.interval) clearInterval(this.interval)
    if (this.options.reloadInterval && this.options.environmentId) {
      this.interval = setInterval(() => this.reload(), this.options.reloadInterval)
    }
  }

  reload () {
    const { backendConnector, languageUtils, logger } = this.services || { logger: console }
    if (!backendConnector) return
    const currentLanguage = backendConnector.language
    if (currentLanguage && currentLanguage.toLowerCase() === 'cimode') return // avoid loading resources for cimode

    const toLoad = []
    const append = (lng) => {
      const lngs = languageUtils.toResolveHierarchy(lng)
      lngs.forEach(l => {
        if (toLoad.indexOf(l) < 0) toLoad.push(l)
      })
    }

    append(currentLanguage)

    if (this.allOptions.preload) this.allOptions.preload.forEach((l) => append(l))

    toLoad.forEach(lng => {
      this.allOptions.ns.forEach(ns => {
        backendConnector.read(lng, ns, 'read', null, null, (err, data) => {
          if (err) logger.warn(`loading namespace ${ns} for language ${lng} failed`, err)
          if (!err && data) logger.log(`loaded namespace ${ns} for language ${lng}`, data)

          backendConnector.loaded(`${lng}|${ns}`, err, data)
        })
      })
    })
  }

  getLanguages (callback) {
    let deferred
    if (!callback) {
      deferred = defer()
      callback = (err, ret) => {
        if (err) return deferred.reject(err)
        deferred.resolve(ret)
      }
    }

    // there are scenarios/users that seems to call getLanguges a lot of time
    this.getLanguagesCalls = this.getLanguagesCalls || []
    this.getLanguagesCalls.push(callback)
    if (this.getLanguagesCalls.length > 1) return

    this.loadUrl({}, getAvailableLanguages(), (err, ret, info) => {
      if (!this.somethingLoaded && info && info.resourceNotExisting) {
        const e = new Error(`Fluentc environment ${this.options.environmentId} does not exist!`)
        const clbs = this.getLanguagesCalls
        this.getLanguagesCalls = []
        return clbs.forEach((clb) => clb(e))
      }

      console.log(ret)

      this.somethingLoaded = true
      const clbs = this.getLanguagesCalls
      this.getLanguagesCalls = []
      clbs.forEach((clb) => clb(err, ret))
    })
    return deferred
  }

  read (language, namespace, callback) {
    const { logger } = this.services || { logger: console }

    const isMissing = isMissingOption(this.options, ['environmentId'])
    if (isMissing) return callback(new Error(isMissing), false)

    this.loadUrl({}, getContent(this.options.environmentId, language), (err, ret, info) => {
      if (!this.somethingLoaded) {
        if (info && info.resourceNotExisting) {
          logger.error('Environment not existing');
          // setTimeout(() => this.checkIfProjectExists(), this.options.checkForProjectTimeout)
        } else {
          this.somethingLoaded = true
        }
      }
      const objectRet = {};
      if (ret && ret.length) {
        for (let i = 0; i < ret.length; i++) {
          objectRet[ret[i].key] = ret[i].value
        }
      }
      callback(err, objectRet)
    })
  }

  loadUrl (options, payload, callback) {
    options = defaults(options, this.options)
    if (typeof payload === 'function') {
      callback = payload
      payload = undefined
    }
    callback = callback || (() => {})
    request(options, this.options.apiPath, payload, (err, res) => {
      const resourceNotExisting = res && res.resourceNotExisting

      if (res && (res.status === 408 || res.status === 400)) { // extras for timeouts on cloudfront
        return callback('failed loading ' + payload.type, false /* retry */, { resourceNotExisting })
      }
      if (res && ((res.status >= 500 && res.status < 600) || !res.status)) { return callback('failed loading ' + payload.type, false /* retry */, { resourceNotExisting }) }
      if (res && res.status >= 400 && res.status < 500) { return callback('failed loading ' + payload.type, false /* no retry */, { resourceNotExisting }) }
      if (!res && err && err.message && err.message.indexOf('Failed to fetch') > -1) { return callback('failed loading ' + payload.type, false /* retry */, { resourceNotExisting }) }
      if (err) return callback(err, false)

      let ret, parseErr
      try {
        ret = JSON.parse(res.data)
        ret = ret.data[payload.type].body
      } catch (e) {
        parseErr = 'failed parsing ' + payload.type + ' to json'
      }
      if (parseErr) return callback(parseErr, false)
      if (this.options.failLoadingOnEmptyJSON && !Object.keys(ret).length) { return callback('loaded result empty for ' + payload.type, false, { resourceNotExisting }) }
      callback(null, ret, { resourceNotExisting })
    })
  }
}

Backend.type = 'backend'

export default Backend
